src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,408 @@
+#include "split_and_merge_algorithm.h"
+#include <iostream>
+#include <algorithm>
+#include <random>
+#include <cmath>
+#include "src/utils/distributions.h"
+
+void SplitAndMergeAlgorithm::read_params_from_proto(
+  const bayesmix::AlgorithmParams &params){
+  BaseAlgorithm::read_params_from_proto(params);
+  T = params.splitmerge_n_restr_gs_updates();
+  K = params.splitmerge_n_mh_updates();
+  M = params.splitmerge_n_full_gs_updates();
+} 
+
+void SplitAndMergeAlgorithm::initialize(){
+  MarginalAlgorithm::initialize();
+
+  if(mixing->get_id()!=bayesmix::MixingId::DP){
+    throw std::invalid_argument(
+      "Invalid mixing supplied to Split and Merge, only DP mixing supported");
+  }
+}
+
+void SplitAndMergeAlgorithm::print_startup_message() const {
+  std::string msg = "Running Split and Merge algorithm with " +
+                    bayesmix::HierarchyId_Name(unique_values[0]->get_id()) +
+                    " hierarchies, " +
+                    bayesmix::MixingId_Name(mixing->get_id()) + " mixing...";
+  std::cout << msg << std::endl;
+}
+
+void SplitAndMergeAlgorithm::sample_allocations(){
+  auto &rng = bayesmix::Rng::Instance().get();
+
+  // MH updates
+  for(unsigned int k=0; k<K; ++k){
+    unsigned int n_data = data.rows();
+
+    // Sample i and j from the datapoints
+    Eigen::VectorXd probas = Eigen::VectorXd::Constant(n_data, 1.0/n_data);
+    unsigned int i = bayesmix::categorical_rng(probas, rng, 0);
+    probas = Eigen::VectorXd::Constant(n_data-1, 1.0/(n_data-1));
+    unsigned int j = bayesmix::categorical_rng(probas, rng, 0);
+    if(j>=i){
+      j++;
+    }
+
+    compute_S(i,j);
+    compute_C_launch(i,j);
+
+    // Restricted GS steps
+    for(unsigned int t=0; t<T; ++t){
+      restricted_GS(i,j);
+    }
```

Comment:
I would give more explicative names to all of these methods

Reply:
Concerning the function `restricted_GS`, the problem is in the term "restricted Gibbs Sampling" or in the absence of information about what the function does?
More precisely, should we find a more explicative substitute for "restricted Gibbs Sampling" or should we enhance the name of the function to make it more explicative (e.g. something like `restricted_GS_step` or `sample_restricted_GS`)?

Reply:
I think that restricted_gibbs_sampling would be fine, provided that you write a meaningful docstring about what this function is actually doing in the .h file